pkg/builtin/builtin.go

```diff
@@ -251,13 +265,48 @@ func SysWrite(ctx context.Context, env []string, input string) (string, error) {
 		return "", err
 	}
 
+	// Lock the file to prevent concurrent writes from other tool calls.
+	locker.Lock(params.Filename)
+	defer locker.Unlock(params.Filename)
+
 	data := []byte(params.Content)
 	msg := fmt.Sprintf("Wrote %d bytes to file %s", len(data), params.Filename)
 	log.Debugf(msg)
 
 	return "", os.WriteFile(params.Filename, data, 0644)
 }
 
+func SysAppend(ctx context.Context, env []string, input string) (string, error) {
+	var params struct {
+		Filename string `json:"filename,omitempty"`
+		Content  string `json:"content,omitempty"`
+	}
+	if err := json.Unmarshal([]byte(input), &params); err != nil {
+		return "", err
+	}
+
+	// Lock the file to prevent concurrent writes from other tool calls.
+	locker.Lock(params.Filename)
+	defer locker.Unlock(params.Filename)
+
+	f, err := os.OpenFile(params.Filename, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return "", err
+	}
+
+	// Attempt to append to the file and close it immediately.
+	// Write is guaranteed to return an error when n != len([]byte(params.Content))
+	n, err := f.Write([]byte(params.Content))
+	if err := errors.Join(err, f.Close()); err != nil {
+		return "", err
+	}
+
+	msg := fmt.Sprintf("Appended %d bytes to file %s", n, params.Filename)
+	log.Debugf(msg)
```

Comment:
Why fmt the string first? Not that it matters but by  passing to log.Debugf(x, ...) it will not take the overhead of string formatting when debug logging is disabled.

Reply:
Ah, yup. I just copied this over from SysWrite without thinking about it. I'll fix it.

Reply:
done